### PR TITLE
[DELETED] ipq40xx: Add ZTE MF289F

### PR DESCRIPTION
**Pull request by stich86**
It's a 4G Cat.20 router used by Vodafone Italy (called Vodafone FWA) and Vodafone DE\T-Mobile PL (called GigaCube). Modem is a MiniPCIe-to-USB based on Snapdragon X24, it supports 4CA aggregation.

There are currently two hardware revisions, which differ on the 5Ghz radio:

AT1 = QCA9984 5Ghz Radio on PCI-E bus
AT2 = IPQ4019 5Ghz Radio insde IPQ4019 like 2.4Ghz

Device specification
--------------------
SoC Type: Qualcomm IPQ4019
RAM: 256 MiB
Flash: 128 MiB SPI NAND (Winbond W25N01GV)
ROM: 2MiB SPI Flash (GD25Q16)
Wireless 2.4 GHz (IP4019): b/g/n, 2x2
Wireless 5 GHz (QCA9984): a/n/ac, 4x4 HW REV AT1 OR 5 GHz (IPA4019): a/n/ac, 2x2  HW REV AT2 Ethernet: 2xGbE (WAN/LAN1, LAN2)
USB ports: No
Button: 2 (Reset/WPS)
LEDs: 3 external leds: Network (white or red), Wifi, Power and 1 internal (blue) Power: 12 VDC, 1 A
Connector type: Barrel
Bootloader: U-Boot

Installation
------------
1. Place OpenWrt initramfs image for the device on a TFTP in the server's root. This example uses Server IP: 192.168.0.2
2. Connect serial console (115200,8n1) to serial connector (refer to this image: https://ibb.co/31Gngpr).
3. Connect TFTP server to RJ-45 port (WAN/LAN1).
4. Stop in u-Boot (using ESC button) and run u-Boot commands:

setenv serverip 192.168.0.2
setenv ipaddr 192.168.0.1
set fdt_high 0x85000000
tftp openwrt-ipq40xx-generic-zte_mf289f-initramfs-fit-zImage.itb bootm $loadaddr

5. Please make backup of original partitions, if you think about revert to stock, specially mtd16 (Web UI) and mtd17 (rootFS). Use /tmp as temporary storage and do:

WEB PARITION
--------------------------------------
cat /dev/mtd16 > /tmp/mtd16.bin
scp /tmp/mtd16.bin root@YOURSERVERIP:/
rm /tmp/mtd16.bin

ROOT PARITION
--------------------------------------
cat /dev/mtd17 > /tmp/mtd17.bin
scp /tmp/mtd17.bin root@YOURSERVERIP:/
rm /tmp/mtd17.bin

6. Login via ssh or serial and remove stock partitions (default IP 192.168.0.1):

ubiattach -m 17 # it could return error if ubi was attached before or rootfs part was erased before ubirmvol /dev/ubi0 -N ubi_rootfs # it could return error if rootfs part was erased before ubirmvol /dev/ubi0 -N ubi_rootfs_data # some devices doesn't have it

7. Install image via “sysupgrade -n” (like “sysupgrade -n /tmp/openwrt-ipq40xx-generic-zte_mf289f-squashfs-sysupgrade.bin”; previously wgeting bin). Sometimes it could print ubi attach error, but please ignore it if process goes forward.

Flash Layout

NAND:
mtd8: 000a0000 00020000 "fota-flag"
mtd9: 00080000 00020000 "0:ART"
mtd10: 00080000 00020000 "mac"
mtd11: 000c0000 00020000 "reserved2"
mtd12: 00400000 00020000 "cfg-param"
mtd13: 00400000 00020000 "log"
mtd14: 000a0000 00020000 "oops"
mtd15: 00500000 00020000 "reserved3"
mtd16: 00800000 00020000 "web"
mtd17: 01d00000 00020000 "rootfs"
mtd18: 01900000 00020000 "data"
mtd19: 03200000 00020000 "fota"
mtd20: 0041e000 0001f000 "kernel"
mtd21: 0101b000 0001f000 "ubi_rootfs"

SPI:
mtd0: 00040000 00010000 "0:SBL1"
mtd1: 00020000 00010000 "0:MIBIB"
mtd2: 00060000 00010000 "0:QSEE"
mtd3: 00010000 00010000 "0:CDT"
mtd4: 00010000 00010000 "0:DDRPARAMS"
mtd5: 00010000 00010000 "0:APPSBLENV"
mtd6: 000c0000 00010000 "0:APPSBL"
mtd7: 00050000 00010000 "0:reserved1"

Back to Stock (!!! need original dump taken from initramfs !!!) -------------
1. Place mtd16.bin and mtd17.bin  initramfs image for the device on a TFTP in the server's root. This example uses Server IP: 192.168.0.2
2. Connect serial console (115200,8n1) to serial console connector (refer to this image: https://ibb.co/31Gngpr).
3. Connect TFTP server to RJ-45 port (WAN/LAN1).
4. rename mtd16.bin to web.img and mtd17.bin to root_uImage_s
5. Stop in u-Boot (using ESC button) and run u-Boot commands:

This will erase RootFS+Web:
nand erase 0x1000000 0x800000
nand erase 0x1800000 0x1D00000

This will restore RootFS:
tftpboot 0x84000000 ${dir}root_uImage_s
nand erase 0x1800000 0x1D00000
nand write $fileaddr 0x1800000 $filesize

This will restore Web Interface:
tftpboot 0x84000000 ${dir}web.img
nand erase 0x1000000 0x800000
nand write $fileaddr 0x1000000 $filesize

After first boot on stock firwmare, do a factory reset. Push reset button for 5 seconds so all parameters will be reverted to the one printed on label on bottom of the router

### CODE REVIEW

File: package/firmware/ipq-wifi/Makefile [hunk 1]
```diff
@@ -29,15 +29,14 @@ ALLWIFIBOARDS:= \
 	devolo_magic-2-wifi-next \
 	edgecore_ecw5410 \
 	edgecore_oap100 \
-	extreme-networks_ws-ap3915i \
```
Comment (obsy) [inline package/firmware/ipq-wifi/Makefile:32]: not part of this commit?

File: package/firmware/ipq-wifi/Makefile [hunk 2]
```diff
@@ -29,15 +29,14 @@ ALLWIFIBOARDS:= \
...
 	glinet_gl-ap1300 \
 	glinet_gl-s1300 \
 	linksys_ea8300 \
 	p2w_r619ac \
-	pakedge_wr-1 \
```
Comment (obsy) [inline package/firmware/ipq-wifi/Makefile:37]: not part of this commit?

File: package/firmware/ipq-wifi/Makefile [hunk 3]
```diff
@@ -102,14 +101,13 @@ $(eval $(call generate-ipq-wifi-package,aruba_ap-365,Aruba AP-365))
 $(eval $(call generate-ipq-wifi-package,devolo_magic-2-wifi-next,devolo Magic 2 WiFi next))
 $(eval $(call generate-ipq-wifi-package,edgecore_ecw5410,Edgecore ECW5410))
 $(eval $(call generate-ipq-wifi-package,edgecore_oap100,Edgecore OAP100))
-$(eval $(call generate-ipq-wifi-package,extreme-networks_ws-ap3915i,Edgecore OAP100))
```
Comment (obsy) [inline package/firmware/ipq-wifi/Makefile:105]: not part of this commit?

File: package/firmware/ipq-wifi/Makefile [hunk 4]
```diff
@@ -102,14 +101,13 @@ $(eval $(call generate-ipq-wifi-package,aruba_ap-365,Aruba AP-365))
...
 $(eval $(call generate-ipq-wifi-package,glinet_gl-ap1300,GL.iNet GL-AP1300))
 $(eval $(call generate-ipq-wifi-package,glinet_gl-s1300,GL.iNet GL-S1300))
 $(eval $(call generate-ipq-wifi-package,linksys_ea8300,Linksys EA8300))
 $(eval $(call generate-ipq-wifi-package,p2w_r619ac,P&W R619AC))
-$(eval $(call generate-ipq-wifi-package,pakedge_wr-1,Pakedge WR-1))
```
Comment (obsy) [inline package/firmware/ipq-wifi/Makefile:110]: not part of this commit?

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 5]
```diff
@@ -138,11 +137,6 @@ ipq40xx_setup_interfaces()
 			"0u@eth0" "2:lan" "3:lan" "4:lan"
 		ucidef_set_interface_wan "eth1"
 		;;
-	pakedge,wr-1)
```
Comment (obsy) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network:141]: not part of this commit?

File: target/linux/ipq40xx/base-files/etc/board.d/02_network [hunk 6]
```diff
@@ -28,7 +28,6 @@ ipq40xx_setup_interfaces()
 	dlink,dap-2610 |\
 	engenius,eap1300|\
 	engenius,emd1|\
-	extreme-networks,ws-ap3915i|\
```
Comment (obsy) [inline target/linux/ipq40xx/base-files/etc/board.d/02_network:31]: not part of this commit?

**Comment by stich86** [comment]
repush with correct files..

File: package/firmware/ipq-wifi/Makefile (hunk 1 above)
Comment (stich86) [reply package/firmware/ipq-wifi/Makefile]: sorry i've updated my local patch using new file, it's fixed now

File: package/firmware/ipq-wifi/Makefile (hunk 3 above)
Comment (stich86) [reply package/firmware/ipq-wifi/Makefile]: sorry i've updated my local patch using new file, it's fixed now

File: target/linux/ipq40xx/base-files/etc/board.d/02_network (hunk 5 above)
Comment (stich86) [reply target/linux/ipq40xx/base-files/etc/board.d/02_network]: sorry i've updated my local patch using new file, it's fixed now

File: target/linux/ipq40xx/base-files/etc/board.d/02_network (hunk 6 above)
Comment (stich86) [reply target/linux/ipq40xx/base-files/etc/board.d/02_network]: sorry i've updated my local patch using new file, it's fixed now

File: package/firmware/ipq-wifi/Makefile (hunk 2 above)
Comment (stich86) [reply package/firmware/ipq-wifi/Makefile:37]: sorry i've updated my local patch using new file, it's fixed now

File: package/firmware/ipq-wifi/Makefile (hunk 4 above)
Comment (stich86) [reply package/firmware/ipq-wifi/Makefile]: sorry i've updated my local patch using new file, it's fixed now

File: package/firmware/ipq-wifi/Makefile [hunk 7]
```diff
@@ -37,7 +37,9 @@ ALLWIFIBOARDS:= \
 	pakedge_wr-1 \
 	qxwlan_e2600ac-c1 \
 	qxwlan_e2600ac-c2 \
-	teltonika_rutx
+	teltonika_rutx \
+        zte_mf289f
+
```
Comment (DragonBluep) [inline package/firmware/ipq-wifi/Makefile:42]: Please drop new line 42, and also use tabs instead of spaces in all files. I see Makefile, 02_network and genneric.mk are still align with spaces.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+	status = "okay";
+};
+
+&tlmm {
+	i2c_0_pins: i2c_0_pinmux {
+			mux {
+				pins = "gpio20", "gpio21";
+				function = "blsp_i2c0";
+				bias-disable;
+			};
```
Comment (DragonBluep) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:333]: Seems there is one more tab in line 329-330, please also notice line 336-374, 419-424.

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+
+	/* This node is used to restart modem module to avoid anomalous behaviours on initial communication  */
+	gpio-restart {
+		compatible = "gpio-restart";
+		gpios = <&tlmm 8 GPIO_ACTIVE_HIGH>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+		led_status: led-status {
```
Comment (DragonBluep) [inline target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:37]: Need new line here.  And I see whole wifi led has been droped? Sorry I didn't make it clear earlier, we just need to remove the `led_wlan: ` string. f4f22098f390d6dc62989bcc8a65400e90cb2695

Comment (stich86) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:37]: what do you mean for "New line"

it should be:

```
leds {

		compatible = "gpio-leds";
		led_status: led-status {
```

instead of

```
leds {
		compatible = "gpio-leds";
		led_status: led-status {
```

Also other stuff doesn't have new line after first parenthesis. Do I need to add on the whole DTS?
Don't know why all this errors, base DTS was from ZTE-MF286D, except for the tabs the structure is pretty similar

Let me know, just to avoid too many resquash, thx :)

Comment (stich86) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:37]: > Need new line here. And I see whole wifi led has been droped? Sorry I didn't make it clear earlier, we just need to remove the `led_wlan: ` string. [f4f2209](https://github.com/openwrt/openwrt/commit/f4f22098f390d6dc62989bcc8a65400e90cb2695)

regarding led_wlan, do you mean in this way:

```
                led-wlan {
                        function = LED_FUNCTION_WLAN;
                        color = <LED_COLOR_ID_BLUE>;
                        gpios = <&tlmm 61 GPIO_ACTIVE_LOW>;
                        linux,default-trigger = "phy0tpt";
                };
```

?

thx!

File: package/firmware/ipq-wifi/Makefile (hunk 7 above)
Comment (stich86) [reply package/firmware/ipq-wifi/Makefile:42]: fixed, i'll publish in the next commit when we agree on DTS comment about leds

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts (hunk 8 above)
Comment (stich86) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:333]: fixed, i'll publish in the next commit when we agree on DTS comment about leds

File: target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts (hunk 9 above)
Comment (DragonBluep) [reply target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-mf289f.dts:37]: should be 
```
leds {
		compatible = "gpio-leds";

		led_status: led-status {
```

> regarding led_wlan, do you mean in this way:
> 
> ```
>                 led-wlan {
>                         function = LED_FUNCTION_WLAN;
>                         color = <LED_COLOR_ID_BLUE>;
>                         gpios = <&tlmm 61 GPIO_ACTIVE_LOW>;
>                         linux,default-trigger = "phy0tpt";
>                 };
> ```
> 
> ?
> 
> thx!

Yes, and please also notice space problem. Hope you don't feel I'm too verbose, but some main maintainers may be strict about formatting.

**Comment by stich86** [comment]
resquash..

**Comment by stich86** [comment]
@DragonBluep i've accidentally removed my repo from git...
so i've refork and push using new PR

if it's possible please delete this one

new one here: https://github.com/openwrt/openwrt/pull/10710
